### ISSUE TITLE
Build RAG encoder-classifier for QDE

### ISSUE THREAD

**Issue by OmriDan** [OmriDan/NLP_final_project#5]
# LLM Recommendations

For your question difficulty estimation task:

## Encoder Models (Recommended for Classification):

**DeBERTa-v3 (microsoft/deberta-v3-base):** Excellent performance on classification tasks with reasonable compute requirements
**RoBERTa (roberta-base):** Strong performance and efficiency
**BERT (bert-base-uncased):** Good baseline with lower computational requirements

## Embedding Models (for RAG):

**all-MiniLM-L6-v2:** Fast, lightweight embeddings for retrieval
**all-mpnet-base-v2:** Better quality embeddings with moderate compute
**bge-large-en-v1.5:** State-of-the-art retrieval performance

## LLMs (if you need generation aspects):

**Mistral-7B:** Excellent performance/size ratio
**Phi-2:** Lightweight but capable
**Llama-3-8B:** Strong performance with reasonable compute

You can also use a hybrid approach:

Use encoder models for classification
Use an LLM to generate natural language explanations about difficulty

### Feature Engineering: Consider extracting features like:

Linguistic complexity metrics (readability scores)
Question/answer length ratio
Domain-specific keyword presence
Syntactic complexity measures

### Training Techniques:

Implement label smoothing to account for subjectivity in difficulty ratings
Use curriculum learning (train on easy examples first)
Consider multi-task learning if you have related tasks

### Evaluation:

Use stratified cross-validation due to potential class imbalance
Consider ordinal metrics since difficulty levels have a natural ordering

**Comment by OmriDan** [comment]
# Data for RAG

It should be in the following format:

question,answer,difficulty
"What is the capital of France?","Paris is the capital of France.",0
"Explain the concept of gravitational waves.","Gravitational waves are disturbances in the curvature of spacetime generated by accelerated masses that propagate as waves at the speed of light. They were predicted by Einstein's theory of general relativity and directly detected for the first time in 2015 by LIGO.",2
"What is photosynthesis?","Photosynthesis is the process by which green plants and some other organisms use sunlight to synthesize foods with carbon dioxide and water. It involves the conversion of light energy into chemical energy.",1

The difficulty can be 1-5, 0-1, 1-3 etc.
Ensure your data is balanced across difficulty levels or apply weighting during training

# Knowledge Corpus Format
The knowledge corpus (knowledge_corpus.txt) should contain reference materials that provide context for your questions. Each line should represent a separate document or paragraph:
The capital of France is Paris. Paris is located in the north-central part of France and is the country's largest city with a population of over 2 million people. It is known for landmarks such as the Eiffel Tower, the Louvre Museum, and Notre-Dame Cathedral.

Albert Einstein predicted the existence of gravitational waves in 1916 as part of his general theory of relativity. Gravitational waves are ripples in spacetime caused by some of the most violent and energetic processes in the Universe. They were directly detected for the first time in 2015 by the Laser Interferometer Gravitational-Wave Observatory (LIGO).

Photosynthesis is the process used by plants, algae and certain bacteria to convert light energy into chemical energy. This chemical energy is stored in carbohydrate molecules, such as sugars, which are synthesized from carbon dioxide and water. Oxygen is released as a waste product. The process takes place in the chloroplasts of plant cells, specifically using chlorophyll, the green pigment involved in photosynthesis.

## Key points:

Each line should be a self-contained document or paragraph.
The corpus should cover topics relevant to your questions.
Include a mix of basic and advanced content for different difficulty levels.
The more comprehensive your corpus, the better your RAG system will perform.
Consider including examples of questions and answers from each difficulty level.

**Comment by OmriDan** [comment]
# Prompt Usage:
Unlike pure LLM approaches, this RAG-based classifier **doesn't use prompts during training since it's using a classification head.**

**Comment by OmriDan** [comment]
Used MAE for choosing the best model - the output just got the mean score almost every time.

Possible solution is the over-simplifide one - MLP and sigmoid after getting the encoder layer.

We used additional huggingface datasets to enrich our RAG.Possible test -changing its corpura and looking at model's performance.

Should check this for the regression head:
https://lajavaness.medium.com/regression-with-text-input-using-bert-and-transformers-71c155034b13

**Comment by OmriDan** [comment]
Comparing fine-tuned model VS pretrained (same datasets, same models):

### Fine Tuned
Question: Print Hello World in Python.
Answer: print('Hello World')
Difficulty Score: 0.16
Question: Write a function to find the maximum subarray sum.
Answer: def max_subarray_sum(arr): return max(sum(arr[i:j]) for i in range(len(arr)) for j in range(i+1, len(arr)+1))
Difficulty Score: 0.38
Question: Implement a function to check if a string is a palindrome.
Answer: def is_palindrome(s): return s == s[::-1]
Difficulty Score: 0.12
Question: Create a function to find the nth Fibonacci number using dynamic programming.
Answer: def fibonacci(n): if n <= 1: return n; return fibonacci(n-1) + fibonacci(n-2)
Difficulty Score: 0.21
Question: Implement a depth-first search algorithm for a graph.
Answer: def dfs(graph, start): visited = set(); stack = [start]; while stack: vertex = stack.pop(); if vertex not in visited: visited.add(vertex); stack.extend(set(graph[vertex]) - visited)
Difficulty Score: 0.36

### Pre-trained

Question: Print Hello World in Python.
Predicted difficulty: 0.4169

Question: Write a function to find the maximum subarray sum.
Predicted difficulty: 0.4208

Question: Implement a function to check if a string is a palindrome.
Predicted difficulty: 0.4256

Question: Create a function to find the nth Fibonacci number using dynamic programming.
Predicted difficulty: 0.4244

Question: Implement a depth-first search algorithm for a graph.
Predicted difficulty: 0.4223

### Pre-trained + additional prompt that explaines the task:
Probably didn't work because of the regressor head and the lack of fine-tuning...
Question: Print Hello World in Python.
Predicted difficulty: 0.5131

Question: Write a function to find the maximum subarray sum.
Predicted difficulty: 0.5240

Question: Implement a function to check if a string is a palindrome.
Predicted difficulty: 0.5297

Question: Create a function to find the nth Fibonacci number using dynamic programming.
Predicted difficulty: 0.5202

Question: Implement a depth-first search algorithm for a graph.
Predicted difficulty: 0.5272

### Pre-trained + additional prompt that explaines the task + examples (few-shot):
Question: Print Hello World in Python.
Predicted difficulty: 0.4227

Question: Write a function to find the maximum subarray sum.
Predicted difficulty: 0.4256

Question: Implement a function to check if a string is a palindrome.
Predicted difficulty: 0.4261

Question: Create a function to find the nth Fibonacci number using dynamic programming.
Predicted difficulty: 0.4224

Question: Implement a depth-first search algorithm for a graph.
Predicted difficulty: 0.4224


Conclusion - the regressor module might effect the functionality of the pre-trained models. Therefore the
